leader-election-impl/src/test/java/com/palantir/paxos/PaxosConsensusFastTest.java

```diff
@@ -189,5 +186,67 @@ public void learnerRecovery() {
         cache.clear();
         state.gainLeadership(0);
     }
+
+    @Test
+    public void loseLeadershipAfterSteppingDown() {
+        LeadershipToken token = state.gainLeadership(0);
+        assertThat(state.leader(0).isStillLeading(token)).isEqualTo(StillLeadingStatus.LEADING);
+        assertThat(state.leader(0).stepDown()).isTrue();
+        assertThat(state.leader(0).isStillLeading(token)).isEqualTo(StillLeadingStatus.NOT_LEADING);
+    }
+
+    @Test
+    public void otherNodeCanBecomeLeaderAfterSteppingDown() {
+        LeadershipToken token = state.gainLeadership(0);
+        assertThat(state.leader(0).isStillLeading(token)).isEqualTo(StillLeadingStatus.LEADING);
+        assertThat(state.leader(0).stepDown()).isTrue();
+
+        LeadershipToken token2 = state.gainLeadership(1);
+        assertThat(state.leader(1).isStillLeading(token2)).isEqualTo(StillLeadingStatus.LEADING);
+    }
+
+    @Test
+    public void leadershipIfRegainedImmediatelyIsOnDifferentToken() {
+        LeadershipToken token1 = state.gainLeadership(0);
+        assertThat(state.leader(0).isStillLeading(token1)).isEqualTo(StillLeadingStatus.LEADING);
+        assertThat(state.leader(0).stepDown()).isTrue();
+
+        LeadershipToken token2 = state.gainLeadership(0);
+        assertThat(state.leader(0).isStillLeading(token1)).isEqualTo(StillLeadingStatus.NOT_LEADING);
+        assertThat(state.leader(0).isStillLeading(token2)).isEqualTo(StillLeadingStatus.LEADING);
+        assertThat(state.leader(0).stepDown()).isTrue();
+    }
+
+    @Test
+    public void nonLeaderStepDownDoesNotAffectLeader() {
+        LeadershipToken token = state.gainLeadership(0);
+        assertThat(state.leader(0).isStillLeading(token)).isEqualTo(StillLeadingStatus.LEADING);
+        assertThat(state.leader(1).stepDown()).isFalse();
+        assertThat(state.leader(0).isStillLeading(token)).isEqualTo(StillLeadingStatus.LEADING);
+    }
+
+    @Test
+    public void failToStepDownIfNoQuorum() {
+        LeadershipToken token = state.gainLeadership(0);
+        assertThat(state.leader(0).isStillLeading(token)).isEqualTo(StillLeadingStatus.LEADING);
+
+        knockOutQuorumNotIncludingZero();
+        assertThat(state.leader(0).stepDown()).isFalse();
+
```

Comment:
wouldn't this fail at the proposer step? why do you need to have a special condition inside PLES?

Reply:
This fails actually because you realise you're not the leader. Failing at the proposer bit would throw an exception - this was unfortunately a bit trickier to test, because it relies on a node coming down after you've confirmed you're the leader but before you initiate the Paxos instance.

As for the leadership check (I'm assuming that's the condition we're referring to) - that's because I didn't want a call to `stepDown()` on a non-leader to cause a leadership change in general.